quickwit-actors/src/async_actor.rs

```diff
@@ -0,0 +1,160 @@
+use crate::actor::ActorTermination;
+use crate::actor_handle::ActorHandle;
+use crate::actor_state::ActorState;
+use crate::mailbox::{create_mailbox, Command, Inbox};
+use crate::scheduler::SchedulerMessage;
+use crate::{Actor, ActorContext, KillSwitch, Mailbox, ReceptionResult};
+use anyhow::Context;
+use async_trait::async_trait;
+use tokio::sync::watch::{self, Sender};
+use tracing::{debug, error};
+
+/// An async actor is executed on a regular tokio task.
+///
+/// It can make async calls, but it should not block.
+/// Actors doing CPU heavy work should implement `SyncActor` instead.
+#[async_trait]
+pub trait AsyncActor: Actor + Sized {
+    /// Initialize is called before running the actor.
+    ///
+    /// This function is useful for instance to schedule an initial message in a looping
+    /// actor.
+    ///
+    /// It can is treated just as if it was an initial implicit Initial message.
+    /// Returning an ActorTermination will therefore have the same effect as if it
+    /// was in `process_message`. (e.g. the actor will stop, the finalize method will be called.
+    /// the kill switch may be activated etc.)
+    async fn initialize(&mut self, _ctx: &ActorContext<Self>) -> Result<(), ActorTermination> {
+        Ok(())
+    }
+
+    /// Processes a message.
+    ///
+    /// If true is returned, the actors will continue processing messages.
+    /// If false is returned, the actor will terminate "gracefully".
+    ///
+    /// If Err is returned, the actor will be killed, as well as all of the actor
+    /// under the same kill switch.
+    async fn process_message(
+        &mut self,
+        message: Self::Message,
+        ctx: &ActorContext<Self>,
+    ) -> Result<(), ActorTermination>;
+
+    async fn finalize(
+        &mut self,
+        _termination: &ActorTermination,
+        _ctx: &ActorContext<Self>,
+    ) -> anyhow::Result<()> {
+        Ok(())
+    }
+}
+
+pub(crate) fn spawn_async_actor<A: AsyncActor>(
+    actor: A,
+    kill_switch: KillSwitch,
+    scheduler_mailbox: Mailbox<SchedulerMessage>,
```

Comment:
is this the universes scheduler mailbox? or are there multiple scheduler?

Reply:
One scheduler per universe